driver-core/src/main/com/mongodb/RequestContext.java

```diff
@@ -0,0 +1,158 @@
+     * passed default value.
+     *
+     * @param <T> an unchecked casted generic for fluent typing convenience
+     * @param key          a lookup key to resolve the value within the context
+     * @param defaultValue a fallback value if key doesn't resolve
+     * @return the value resolved for this key, or the given default if not present
+     */
+    @Nullable
+    default <T> T getOrDefault(Object key, @Nullable T defaultValue) {
+        if (!hasKey(key)) {
+            return defaultValue;
+        }
+        return get(key);
+    }
+
+    /**
+     * Resolve a value given a key within the {@link RequestContext}.
+     *
+     * @param <T> an unchecked casted generic for fluent typing convenience
+     * @param key a lookup key to resolve the value within the context
+     * @return an {@link Optional} of the value for that key.
+     */
+    default <T> Optional<T> getOrEmpty(Object key) {
+        if (hasKey(key)) {
+            return Optional.of(get(key));
+        }
+        return Optional.empty();
+    }
+
+    /**
+     * Return true if a particular key resolves to a value within the {@link RequestContext}.
+     *
+     * @param key a lookup key to test for
+     * @return true if this context contains the given key
+     */
+    boolean hasKey(Object key);
+
+    /**
+     * Return true if the {@link RequestContext} is empty.
+     *
+     * @return true if the {@link RequestContext} is empty.
+     */
+    boolean isEmpty();
+
+    /**
+     * Modifies this instance with the given key and value.
+     * If that key existed in the current RequestContext, its associated
+     * value is replaced.
+     *
+     * @param key   the key to add/update
+     * @param value the value to associate to the key
+     * @throws NullPointerException if either the key or value are null
+     */
+    void put(Object key, Object value);
+
+    /**
+     * Create a new {@link RequestContext} that contains all current key/value pairs plus the
+     * given key/value pair <strong>only if the value is not {@literal null}</strong>. If that key existed in the
+     * current Context, its associated value is replaced in the resulting {@link RequestContext}.
+     *
+     * @param key         the key to add/update in the new {@link RequestContext}
+     * @param valueOrNull the value to associate to the key in the new {@link RequestContext}, null to ignore the operation
+     * @throws NullPointerException if the key is null
+     */
+    default void putNonNull(Object key, @Nullable Object valueOrNull) {
+        if (valueOrNull != null) {
+            put(key, valueOrNull);
+        }
+    }
+
+    /**
+     * Delete the given key and its associated value from the RequestContext.
+     *
+     * @param key the key to remove.
+     */
+    void delete(Object key);
+
+    /**
+     * Return the size of this {@link RequestContext}, the number of immutable key/value pairs stored inside it.
```

Comment:
The method `stream` still does not specify whether the returned `Map.Entry` objects are modifiable or not, and whether modifying such a `Map.Entry` results in modifying the state of the `RequestContext. I propose to specify this, and the best thing is probably to say that `Map.Entry` are unmodifiable (we may further require immutability, I am fine with either option).

Reply:
I don't think the driver particularly cares whether the Map.Entry results are immutable, since the driver never uses this method.  Can implementers just decide for themselves how they want to handle this?  Do we need to restrict it?

Reply:
If we are leaving this up to an implementation, I propose to state this explicitly. This way we show, that this is not an oversight, i.e., there is no rule that we forgot to specify.

Reply:
Agreed.  Done.